## 异步模式

### 回调

- 默认情况下，Jest 测试一旦执行到末尾就会完成，所以

  ```JS
  // 不要这样做！
  test('the data is peanut butter', () => {
    function callback(data) {
      expect(data).toBe('peanut butter');
    }
  
    fetchData(callback);
  });
  
  // 要这样做
  test('the data is peanut butter', done => {
    function callback(data) {
      expect(data).toBe('peanut butter');
      done(); // Jest会等done回调函数执行结束后，结束测试
    }
  
    fetchData(callback);
  });
  ```

### Promise

```JS
test('the data is peanut butter', () => {
  return fetchData().then(data => { // 你忘了 return 语句的话，在 fetchData 返回的这个 promise 被 resolve、then() 有机会执行之前，测试就已经被视为已经完成了
    expect(data).toBe('peanut butter');
  });
});
```

### resolves` / `.rejects

```JS
test('the data is peanut butter', () => {
  return expect(fetchData()).resolves.toBe('peanut butter'); // 有 return
});

test('the fetch fails with an error', () => {
  return expect(fetchData()).rejects.toMatch('error');
});
```

### ！！！Async / Await

```JS
test('the data is peanut butter', async () => {
  const data = await fetchData();
  expect(data).toBe('peanut butter');
});

test('the fetch fails with an error', async () => {
  expect.assertions(1); // 如果你想要 Promise 被拒绝，使用 .catch 方法。 请确保添加 expect.assertions 来验证一定数量的断言被调用。 否则一个fulfilled态的 Promise 不会让测试失败
  try {
    await fetchData();
  } catch (e) {
    expect(e).toMatch('error');
  }
});
```

